### PR TITLE
Add "Duplicate sample", sample copy/paste

### PR DESCRIPTION
**Pull request by defensem3ch**
Implements https://github.com/jkotlinski/lsdpatch/issues/69

### CODE REVIEW

File: src/main/java/kitEditor/KitEditor.java [hunk 1]
```diff
@@ -1132,7 +1140,62 @@ private void trimAllSamples() {
...
+        if (dest != -1) {
+            // copy sample data
+            try {
+            dupeSample = Sample.dupeSample(firstSample);
+            dupeSample.reload(halfSpeed.isSelected());
+            } catch (Exception e) {
+                showFileErrorMessage(e);
+                return;
+            }
+            samples[selectedBank][dest] = dupeSample;
```
Comment (Eiyeron) [inline src/main/java/kitEditor/KitEditor.java:1160]: Codestyle note : this is the kind of changes I'd place at the end of the whole `dupleSample` rename/trim process, just to avoid having to reset `samples[selectedBank][dest]` to `null` if part of the process went wrong. It makes the code more manageable in the future as there'll be less chances that a refactor will leave a half-broken sample in the `samples` array (got fun crashes like this in the past with multi-threaded code).

For future reference, this is the idea of commits in database software, either do all  the changes at once or not, there is no in-betweens.

File: src/main/java/kitEditor/KitEditor.java [hunk 2]
```diff
@@ -1132,7 +1140,62 @@ private void trimAllSamples() {
...
+            renameSample(dest, firstSample.getName());
+            if (bytesFree() < 0 && firstSample.canAdjustVolume()) {
+                int fixedTrim = dupeSample.getTrim() - bytesFree() / 16;
+                assert fixedTrim > 0;
+                dupeSample.setTrim(fixedTrim);
+            } else if (bytesFree() < 0) {
+                samples[selectedBank][dest] = null;
+                JOptionPane.showMessageDialog(contentPane,
```
Comment (Eiyeron) [inline src/main/java/kitEditor/KitEditor.java:1168]: Maybe in the future, do throw exceptions instead to scope the whole algorithm with one `try/except` that'll pop the error message for all kinds of issues that can raise here up at once?

**Review by Eiyeron — COMMENTED** [review]
I'll let Johan approve or not the PR, but I'd just happen to pass by this PR and wanted to look at it a little bit to see how it'd go. Honestly it looks good to me, except the trim part where I don't know enough of the code changes to know what it's about (trimming head/tail to remove parts too silent?).

File: src/main/java/kitEditor/Sample.java [hunk 3]
```diff
@@ -4,9 +4,9 @@
 import java.util.Random;
 import javax.sound.sampled.*;
 
-class Sample {
+class Sample implements Cloneable {
```
Comment (jkotlinski) [inline src/main/java/kitEditor/Sample.java:7]: Cloneable is not implemented in conventional way.
According to Java docs:
https://docs.oracle.com/javase/8/docs/api/java/lang/Cloneable.html
" By convention, classes that implement this interface should override Object.clone (which is protected) with a public method. See Object.clone() for details on overriding this method. "

I think Cloneable and Object.clone() seem complicated and confusing...
It might be more simple and to the point to add a copy constructor.

private Sample(Sample otherSample) {
    // make myself a deep copy of the otherSample
}

What do you think?

File: src/main/java/kitEditor/Sample.java [hunk 4]
```diff
@@ -104,6 +110,10 @@ public static Sample createFromWav(File file,
         return s;
     }
 
+    public static Sample dupeSample(Sample sample) throws IOException, CloneNotSupportedException {
+      return (Sample)sample.clone();
+    }
```
Comment (jkotlinski) [inline src/main/java/kitEditor/Sample.java:115]: This one confuses me.
Is there any reason why one would call `Sample.dupleSample(sample)` instead of `sample.clone()`?

File: src/main/java/kitEditor/KitEditor.java [hunk 5]
```diff
@@ -1132,7 +1140,62 @@ private void trimAllSamples() {
                 "Done",
                 JOptionPane.INFORMATION_MESSAGE);
     }
+
+    private void duplicateSample(Sample sample) {
```
[the remark this answers is not in the document]
Comment (jkotlinski) [reply src/main/java/kitEditor/KitEditor.java:1146]: I'm not sure I understood either what it does
but if it's like you describe, maybe it could be renamed to simply
`private void add(Sample sample);` ?

With some luck, then methods like `addSample(File wavFile)` could use it

File: src/main/java/kitEditor/KitEditor.java [hunk 6]
```diff
@@ -897,6 +905,7 @@ private void addSample(File wavFile) {
     }
 
     private void renameSample(int sampleIndex, String sampleName) {
+        samples[selectedBank][sampleIndex].setName(sampleName);
```
Comment (jkotlinski) [inline src/main/java/kitEditor/KitEditor.java:908]: what is the reason for this change?

File: src/main/java/kitEditor/Sample.java (hunk 4 above)
Comment (jkotlinski) [reply src/main/java/kitEditor/Sample.java:115]: Something to watch out for...
It might be a source of bugs, that Object.clone() only does shallow copy of the object fields.
This is probably not what you want!

File: src/main/java/kitEditor/Sample.java (hunk 3 above)
Comment (defensem3ch) [reply src/main/java/kitEditor/Sample.java:7]:
```
private Sample(Sample otherSample) {
    // make this a deep copy of the otherSample
}
```
This seems sensible and from what little searching I have done it seems copy constructor is recommended over Object.clone().

File: src/main/java/kitEditor/Sample.java (hunk 4 above)
Comment (defensem3ch) [reply src/main/java/kitEditor/Sample.java:115]: I will defer to your judgment on this, and as mentioned earlier it would probably be more sensible to use a copy constructor.

File: src/main/java/kitEditor/KitEditor.java (hunk 6 above)
Comment (defensem3ch) [reply src/main/java/kitEditor/KitEditor.java:908]: `Sample.name` needs to be changed when renaming the sample if a new duplicate sample should retain the name when the original has been previously renamed.

File: src/main/java/kitEditor/Sample.java [hunk 7]
```diff
@@ -27,10 +27,27 @@ public Sample(short[] iBuf, String iName) {
         name = iName;
     }
 
+    Sample(Sample s) {
+        file = s.file;
+        name = s.name;
+        originalSamples = s.originalSamples;
+        processedSamples = s.processedSamples;
```
Comment (jkotlinski) [inline src/main/java/kitEditor/Sample.java:34]: I suspect it would be safer to make copies of file, name, originalSamples and processedSamples, rather than just copy the reference.

Probably it doesn't matter though, for the simple copy/paste scenario

**Review by jkotlinski — APPROVED** [review]
